File: LeetCode/279.perfect-squares.cpp
/*
 * @lc app=leetcode id=279 lang=cpp
 *
 * [279] Perfect Squares
 */

// @lc code=start
class Solution
{
public:
    int numSquares(int n)
    {
        int out = 0;
        vector<int> arr;
        for (int i = 1; i * i <= n; i++)
        {
            arr.push_back(i * i);
        }
        bool visited[10001];
        memset(visited, false, sizeof(visited));
        queue<int> q({n});
        while (!q.empty())
        {
            int size = q.size(), x;
            for (int i = 0; i < size; i++)
            {
                x = q.front();
                q.pop();
                if (x == 0)
                {
                    return out;
                }
                else if (!visited[x])
                {
                    visited[x] = true;
                    for (int j = arr.size() - 1; j >= 0; j--)
                    {
                        if (x >= arr[j] && !visited[x - arr[j]])
                        {
                            q.push(x - arr[j]);
                        }
                    }
                }
            }
            out++;
        }
        return out;
    }
};
// @lc code=end

File: LeetCode/279.perfect-squares.py
#
# @lc app=leetcode id=279 lang=python3
#
# [279] Perfect Squares
#


# @lc code=start
import math


class Solution:
    def numSquares(self, n: int) -> int:
        choices = [x * x for x in range(1, n + 1) if x <= math.sqrt(n)]
        dp = [n] * (n + 1)
        dp[0] = 0

        for target in range(n + 1):
            for choice in choices:
                if target - choice < 0:
                    break

                dp[target] = min(dp[target], 1 + dp[target - choice])

        return dp[n]


# @lc code=end
